src/besselk.jl

```diff
@@ -114,3 +129,46 @@ function besselk1x(x::T) where T <: Union{Float32, Float64}
         return muladd(evalpoly(inv(x), P3_k1(T)), inv(evalpoly(inv(x), Q3_k1(T))), Y2_k1(T)) / sqrt(x)
     end
 end
+#=
+Recurrence is used for all values of nu. If besselk0(x) or besselk1(0) is equal to zero
+this will underflow and always return zero even if besselk(x, nu)
+is larger than the smallest representing floating point value.
+In other words, for large values of x and small to moderate values of nu,
+this routine will underflow to zero.
+For small to medium values of x and large values of nu this will overflow and return Inf.
+
+In the future, a more efficient algorithm for large nu should be incorporated.
+=#
+"""
+    besselk(x::T) where T <: Union{Float32, Float64}
+
+Modified Bessel function of the second kind of order nu, ``K_{nu}(x)``.
+"""
+function besselk(nu::Int, x::T) where T <: Union{Float32, Float64}
+    return three_term_recurrence(x, besselk0(x), besselk1(x), nu)
+end
+"""
+    besselk(x::T) where T <: Union{Float32, Float64}
+
+Scaled modified Bessel function of the second kind of order nu, ``K_{nu}(x)*e^{x}``.
+"""
+function besselkx(nu::Int, x::T) where T <: Union{Float32, Float64}
+    return three_term_recurrence(x, besselk0x(x), besselk1x(x), nu)
+end
+
+@inline function three_term_recurrence(x, k0, k1, nu)
+    nu == 0 && return k0
+    nu == 1 && return k1
+
+    # this prevents us from looping through large values of nu when the loop will always return zero
+    (iszero(k0) || iszero(k1)) && return zero(x) 
+
+    k2 = k0
+    for n in 1:nu-1
+        a = 2 * n / x
+        k2 = muladd(a, k1, k0)
```

Comment:
Will `a*k1>k0` generally hold? If not, this should probably use compensated arithmetic to ensure an accurate answer.

Reply:
It doesn't. when x is large k0 will be bigger.

Reply:
This does seem to matter
```julia
# ArbNumerics
julia> Float64(besselk(500, ArbFloat(100)))
2.7313831719901783e279
# current implementation in BigFloat
julia> Float64(three_term_recurrence(BigFloat(100), BigFloat(k0), BigFloat(k1), 500))
2.7313831719901783e279
# current implementation in Float64
julia> besselk(500, 100.0)
2.731383171990204e279
```
Though I am not sure the best way to compensate for this........

Reply:
Does this work?
```
@inline function three_term_recurrence(x::T, k0, k1, nu) where T<:Number
    nu == 0 && return k0
    nu == 1 && return k1

    # this prevents us from looping through large values of nu when the loop will always return zero
    (iszero(k0) || iszero(k1)) && return zero(x) 

    k2 = k0
    k0lo = k1lo = k2lo = zero(T)
    invx2 = 2 / x
    invx2lo = fma(invx2, -x, 2) / x
    for n in 1:nu-1
        a = invx2 * n
        alo = invx2lo * n
        k2 = a * k1
        k2lo = fma(a, -k1, k2) + muladd(k1, alo, muladd(a, k1lo, k0lo))
        k2new = k2 + k0
        k2lo += k2 - (k2new - k0)
        k2 = k2new
        k0, k0lo = k1, k1lo
        k1, k1lo = k2, k2lo
    end
    return k2 + k2lo
end
```

Reply:
Yes!!!
```julia
julia> three_term_recurrence(100.0, k0, k1, 500)
2.7313831719901783e279
```

Reply:
Though there is a performance hit to doing this...

```julia
## your version
julia> @benchmark besselk(200, x) setup=(x=rand()*1000 + 20.0)
BenchmarkTools.Trial: 2481 samples with 995 evaluations.
 Range (min … max):   27.042 ns … 27.604 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     769.602 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):     2.024 μs ±  5.500 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▆█                                               ▁            
  ██▅▃▁▁▄▁▃▁▁▁▁▃▁▄▃▁▅▁▅▁▁▃▃▁▅▃▄▃▅▁▃▅▃▃▁▃▁▃▃▁▄▁▅▄▃▃▅██▆▅▆▇▆▅▆▅▆ █
  27 ns         Histogram: log(frequency) by time      26.1 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

## old version
julia> @benchmark besselk(200, x) setup=(x=rand()*1000 + 20.0)
BenchmarkTools.Trial: 10000 samples with 995 evaluations.
 Range (min … max):   27.022 ns … 7.930 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     212.084 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   366.389 ns ± 1.151 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

   █                                                           
  ██▁▁▁▂▂▁▂▁▂▁▂▂▁▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂ ▂
  27 ns           Histogram: frequency by time        7.92 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

Reply:
I'm honestly kinda shocked I got the error compensation right. That never happens the first time.

Reply:
Definitely impressed!!